### PR TITLE
OD-414 [Feature] Show a customisable toast message for invalid forms

### PR DESCRIPTION
**Pull request by YaroslavOvdii**
@sofiiakvasnevska 

## Issue
https://weboo.atlassian.net/browse/OD-414 - scenario 1
https://share.getcloudapp.com/NQuw9Ork - scenario 2
https://share.getcloudapp.com/kpuKvBbw - scenario 3
https://share.getcloudapp.com/Blu4PL5m - scenario 4
https://share.getcloudapp.com/Koud8zwD - scenario 5
https://share.getcloudapp.com/8LubemrN scebario 6

## Description
Added new Hook `isFormInvalid` so studio users will have access to the invalid form fields and can work with them with custom code.

## Screenshots/screencasts
https://share.getcloudapp.com/ApuYq6jR

## Backward compatibility
This change is fully backward compatible.

## Reviewers
@upplabs-alex-levchenko @AndrRyaz @ivandevupp

### CODE REVIEW

File: js/libs/form.js [hunk 1]
```diff
@@ -472,185 +474,239 @@ Fliplet.Widget.instance('form-builder', function(data) {
...
+              .then(function(result) {
+                if (!result.length) {
+                  showValidationMessage('');
+
+                  return reject();
+                }
+                
+                return resolve();
+              })
+              .catch(function(responce) {
```
Comment (tonytlwu) [inline js/libs/form.js:521]:
```suggestion
              .catch(function(response) {
```

File: js/libs/form.js [hunk 2]
```diff
@@ -472,185 +474,239 @@ Fliplet.Widget.instance('form-builder', function(data) {
...
+          });
+  
+          // We might use this code to save the form data locally when going away from the page
+          // $(window).unload(function onWindowUnload() {
+          //   localStorage.setItem('fl-form-data-' + data.id, this.fields.map(function (field) {
+          //     return { name: field.name, value: field.value };
+          //   }));
+          // });
+        })
+        .catch(function() {
```
Comment (tonytlwu) [inline js/libs/form.js:707]: Currently, if a form submission fails, e.g. no security permission, does the app user see a message? If so, will this silent `.catch()` end up silencing failed submissions? We don't want that.

**Review by tonytlwu — CHANGES_REQUESTED** [review]
Rejected +1

File: js/libs/form.js [hunk 3]
```diff
@@ -472,185 +474,242 @@ Fliplet.Widget.instance('form-builder', function(data) {
...
-          $vm.isSending = false;
-          return;
+          return Promise.resolve();
         }
 
-        this.fields.forEach(function(field) {
-          var value = field.value;
-          var type = field._type;
+        return onFormSubmission()
+        .then(function() {
```
Comment (tonytlwu) [inline js/libs/form.js:541]: This indentation is wrong. Either:

```suggestion
        return onFormSubmission()
          .then(function() {
```

...or...

```suggestion
        return onFormSubmission().then(function() {
```

Then make sure the rest of this promise chain follows the same pattern.

**Review by tonytlwu — CHANGES_REQUESTED** [review]
Rejected +1

File: js/libs/form.js (hunk 3 above)
Comment (YaroslavOvdii) [reply js/libs/form.js:541]: @tonytlwu Fixed

File: js/libs/form.js [hunk 4]
```diff
@@ -472,185 +474,241 @@ Fliplet.Widget.instance('form-builder', function(data) {
...
+            Fliplet.Hooks.run('onFormSubmitError', { formData: formData, error: err });
+          });
+  
+          // We might use this code to save the form data locally when going away from the page
+          // $(window).unload(function onWindowUnload() {
+          //   localStorage.setItem('fl-form-data-' + data.id, this.fields.map(function (field) {
+          //     return { name: field.name, value: field.value };
+          //   }));
+          // });
+        })
```
Comment (tonytlwu) [inline js/libs/form.js:711]:
```suggestion
        });
```

**Review by tonytlwu — CHANGES_REQUESTED** [review]
Rejected +1

Does not conform to coding standard

File: js/libs/form.js (hunk 4 above)
Comment (YaroslavOvdii) [reply js/libs/form.js]: @tonytlwu Fixed

File: js/libs/form.js (hunk 2 above)
Comment (tonytlwu) [reply js/libs/form.js]: Reported [here](https://weboo.atlassian.net/browse/ID-1009)

**Review by tonytlwu — APPROVED** [review]
@sofiiakvasnevska This is ready for testing

Risk: Critical
Affected: All form components
Verify: Forms can be successfully submitted
Squash and merge: Use PR title

File: js/libs/form.js [hunk 5]
```diff
@@ -472,185 +474,241 @@ Fliplet.Widget.instance('form-builder', function(data) {
...
+         *     and will not show the toast
+         *  3. In case when listener was resolved with Promise.reject() we will not show the toast
+         *     and not allow form to submit
+         *  4. In case when listener was resolved with Promise.reject('') we will show the toast with default text
+         *  5. In case when listener was resolved with Promise.reject('error text') we will show the toast
+         *     with 'error text' message
+         */
+        function onFormInvalid() {
+          return new Promise(function(resolve, reject) {
+            Fliplet.Hooks.run('isFormInvalid', invalidFields)
```
Comment (squallstar) [inline js/libs/form.js:511]: @tonytlwu @YaroslavOvdii I don't think the wrapping promise here was necessary: `run` already returns a promise.

File: js/libs/form.js [hunk 6]
```diff
@@ -472,185 +474,241 @@ Fliplet.Widget.instance('form-builder', function(data) {
...
-          $vm.error = fieldErrors.join('. ');
-          $vm.isSending = false;
-          return;
+          return Promise.resolve();
         }
 
-        this.fields.forEach(function(field) {
-          var value = field.value;
-          var type = field._type;
+        return onFormSubmission().then(function() {
```
Comment (squallstar) [inline js/libs/form.js:540]: Is the `.catch` not handled here? Is this intentional? cc @tonytlwu

File: js/libs/form.js (hunk 5 above)
Comment (tonytlwu) [reply js/libs/form.js:511]: You're right.

File: js/libs/form.js (hunk 6 above)
Comment (tonytlwu) [reply js/libs/form.js:540]: Search for the line where `Fliplet.Hooks.run('onFormSubmitError', { formData: formData, error: err });` is run. I believe that catches the error.

**Pull request merged**